test/sharness/t0252-files-gc.sh

```diff
@@ -42,4 +43,34 @@ test_expect_success "gc okay after adding incomplete node" '
   ipfs refs $ADIR_HASH
 '
 
+MYDIR_HASH=QmRLrxQTdmLZNfDJxFQ9uFrgtj4s6cZU9GzTxCq2YLev24
+HELLO_HASH=QmeV1kwh3333bsnT6YRfdCRrSgUPngKmAhhTa4RrqYPbKT
+
+test_expect_success "add directory with direct pin" '
+  mkdir mydir/ &&
+  echo "hello world!" > mydir/hello.txt &&
+  ipfs add --pin=false -r mydir &&
+  ipfs pin add --recursive=false $MYDIR_HASH &&
+  ipfs cat $HELLO_HASH
+'
+
+test_expect_success "run gc and make sure directory contents are removed" '
+  ipfs repo gc &&
+  test_must_fail ipfs cat $HELLO_HASH
+'
+
+test_expect_success "add incomplete directory and make sure gc is okay" '
+  ipfs files cp /ipfs/$MYDIR_HASH /mydir &&
+  ipfs repo gc
```

Comment:
I would do `ipfs cat $HELLO_HASH` here too, just to be sure the thing is really gone and stays gone.